### ISSUE TITLE
bump dependencies

### ISSUE THREAD

**Issue by CJ-Wright** [regro/cf-scripts#22]
@CJ-Wright commented on [Wed Feb 28 2018](https://github.com/regro/cf-graph/issues/43)

It would be good to have a way to bump dependencies along with the versions.
See: https://github.com/conda/conda-build/blob/master/conda_build/skeletons/pypi.py#L869
@isuruf 
@sodre

**Comment by jakirkham** [comment]
An easier strategy to consider, since we only want to parse the requirements, might be to run `python setup.py egg_info` in the source directory. This will generate some metadata in a `*.egg-info` directory (where `*` is the project name). One of the things it generates in this directory is `requires.txt`. This lists all of the requirements of the package including extra requirements in a standard `requirements.txt` file, which could be loaded and parsed with something like [`requirements-parser`]( https://pypi.python.org/pypi/requirements-parser ) ( seems we don't have that yet https://github.com/conda-forge/staged-recipes/issues/5303 ;).

Also might look at  [pipreqs]( https://github.com/bndr/pipreqs/ ) for some useful thoughts. Though it looks like it mainly scans for imports. Still could be handy for detecting a `setuptools` requirement at run time (e.g. uses `pkg_resources`), which is something people seldom state explicitly. Though maybe it is easier to just find that one piece through "grepping".

**Comment by CJ-Wright** [comment]
How do we decide between run and build requirements?

**Comment by jakirkham** [comment]
They are run dependencies. It doesn't export build dependencies (e.g. `setup_requires`). Just checked.

More generally we have discouraged anyone from adding stuff to build beyond `python` and `setuptools`/`pip` unless it is actually used for the build (e.g. `cython`). There are historic reasons for it, but it is also just cleaner and a bit nicer on the CIs. It will also make this task easier if we stick to that guideline.

IOW if someone actually has a build dependency, it is reasonable to assume that they are linking to it or maybe it's some other build tool. Alternatively it could be an old recipe not following best practices.

IMHO it is fair to say we don't touch build dependencies. It's also probably best as the maintainer may have pinned the build requirements due to to build issues that would not be obvious from inspecting the source (e.g. breaks on some version of `cython`).

Sound reasonable? Other thoughts/concerns?

**Comment by CJ-Wright** [comment]
That sounds reasonable to me. I have seen a few PRs with massive build requirements (as python packages) but those usually end up being not needed.

**Comment by jakirkham** [comment]
Sure. Along those lines it would be reasonable to parse out optional dependencies that maintainers already harden and refresh those dependencies. Am hopeful this ends up being mostly straightforward.

**Comment by jakirkham** [comment]
Linking issue ( https://github.com/pypa/warehouse/issues/474 ) as this would provide an API for querying dependencies from Warehouse.

**Comment by jakirkham** [comment]
This might be more doable now with [Python 3.8's `importlib.metadata`]( https://docs.python.org/3.8/library/importlib.metadata.html ).

**Comment by jakirkham** [comment]
cc @beckermr (as you asked about this in the core meeting earlier 😉)

**Comment by beckermr** [comment]
Thank you!

**Comment by jakirkham** [comment]
Yeah was about to say. I don't think this is a blocker for the auto-merge, but it is highly desirable functionality. 😉

For things like C/C++ this mostly falls out of `conda-forge-pinning`, `run_exports`, migrations, etc. When it doesn't the build already fails, much like what `pip check` would do for us 🙂

**Comment by jakirkham** [comment]
Should add there is some other work along these lines. So hopefully the bot can just leverage this once it's ready.

Please feel free to add more and/or correct me as needed 🙂

cc @ocefpaf @marcelotrevisani

**Comment by marcelotrevisani** [comment]
I hope to speed up the development from my side to also support it, it is one of my goals to have it to the "new skeleton".
Let's see if next week I can present something :)

**Comment by medb** [comment]
It seems that Conda Forge channel has almost all google-cloud-* packages, bust most of them are quite outdated. Are there plans to support then in Conda officially?

**Comment by CJ-Wright** [comment]
@medb what do you mean? This issue is about keeping requirements stated in the `meta.yaml` properly pinned. Is your question related to that? If not please open a new issue!

**Comment by medb** [comment]
Sorry, wrong issue, please delete above comment.

**Comment by jakirkham** [comment]
One option would be to use Grayskull ( https://github.com/regro/cf-scripts/issues/1471 )

**Comment by jakirkham** [comment]
Looks like [PEP 658]( https://peps.python.org/pep-0658/ ) is [now deployed on PyPI]( https://github.com/pypi/warehouse/issues/474#issuecomment-1558492843 )! 🎉

Maybe this is another option for pulling this metadata?

**Comment by flying-sheep** [comment]
There is really nothing in the way of building this right now.
- most packages reliably upload wheels and are therefore supported by the PEP 658 API
- the proposed approach doesn‘t need to work for the rest, as only the package opting into this approach needs to guarantee that it reliably uploads wheels

Since our use case doesn‘t involve generating the whole dependency graph, the vast majority of recipes for PyPI packages could work like this:
1. get PyPI dependencies via PEP 658 (e.g. [like this](https://github.com/flying-sheep/PKGBUILDs/blob/492ebf62099a919bdd247259e97ec3119f832157/src/nvcheck/update/sources/pypi.py#L56-L84))
2. map them to conda-forge dependencies using [cf-graph](https://github.com/regro/cf-graph-countyfair)
3. (optional) apply a set of manual fixes defined in the feedstock in case there are known discrepancies

Also note that by now, best practices for publishing on PyPI are “[trusted publishers](https://docs.pypi.org/trusted-publishers/)” which are *easier* to set up than manually uploading using a PyPI token. So you can expect that most new packages (and well-maintained older packages) will reliably upload wheels and support this approach.

**Comment by jakirkham** [comment]
@flying-sheep are you aware of [Grayskull]( https://conda.github.io/grayskull/ )?

The version update bot has support for using it, but users need to opt-in. Here is an example PR: https://github.com/conda-forge/cloudpickle-feedstock/pull/48

**Comment by flying-sheep** [comment]
It won’t remove dependencies though, right?

**Comment by jakirkham** [comment]
I'm not sure offhand. Maybe Matt would know

That said, this is a tricky problem. As conda-forge packages can depend on C/C++ libraries that wouldn't be listed in `pyproject.toml`, we don't want Grayskull to just remove those libraries

In some future world where [PEP 725]( https://peps.python.org/pep-0725/ ) is adopted, `pyproject.toml`'s could include explicit link outs to external dependencies, which [Grayskull could also map]( https://github.com/conda/grayskull/blob/0e7db2717422fd197f2c7db5689d1009ff05249b/grayskull/strategy/py_toml.py#L113 ). Admittedly there would be a long adoption tail

**Comment by flying-sheep** [comment]
Not necessary. Just calculate the dependency set of both versions and apply the diff.

If anything extra or missing appears in the diff, the automation world mention that to a human, otherwise everything can be done automatically

**Comment by jakirkham** [comment]
AIUI enabling Grayskull updates already includes a hint with that info in the OP of the update PR

Thought you were asking for a fully automated solution

Perhaps it is worth trying first to see how it goes for you. Then we can discuss more concrete feedback based on your experience 🙂
